logstash-core/src/main/java/org/logstash/ackedqueue/FsUtil.java

```diff
@@ -0,0 +1,47 @@
+package org.logstash.ackedqueue;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * File System Utility Methods.
+ */
+public final class FsUtil {
+
+    private FsUtil() {
+    }
+
+    /**
+     * Checks if the request number of bytes of free disk space are available under the given
+     * path.
+     * @param path Directory to check
+     * @param size Bytes of free space requested
+     * @return True iff the
+     * @throws IOException on failure to determine free space for given path's partition
+     */
+    public static boolean hasFreeSpace(final String path, final long size) throws IOException {
+        final File[] partitions = File.listRoots();
+        File location = new File(path).getCanonicalFile();
+        boolean found = false;
+        while (!found) {
+            for (final File partition : partitions) {
+                if (partition.equals(location)) {
+                    found = true;
+                    break;
+                }
+            }
+            if (!found) {
+                location = location.getParentFile();
+                if (location == null) {
+                    throw new IllegalStateException(
```

Comment:
not sure I am following... 🤔  
User uses some strange path -> we throw `new IOException("unable to determine free space for that path")`.??  
And it should be caught exactly where it is supposed be caught, as close to the place where this check is done and **not** bubble up in a more difficult to diagnose context?! 

IMO unchecked exception should generally be avoided and used only for weird runtime problems, not for something that is a **documented** behaviour for which we can trap and deal with?